modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java

```diff
@@ -285,8 +289,10 @@ private class LogEntryAndClosureHandler implements EventHandler<LogEntryAndClosu
         // task list for batch
         private final List<LogEntryAndClosure> tasks = new ArrayList<>(NodeImpl.this.raftOptions.getApplyBatch());
 
+        private @Nullable HybridTimestamp safeTs = null;
```

Comment:
Agree with Roma, to be honest I don't like an idea of inlining safeTime logic inside jraft. Despite the fact that you state opposite idea in PR description "Also it's bad from the pluggable replication point of view, because not all protocols require this concept.".

Reply:
>Safe time doesn't seem to belong to the Raft protocol
I believe you are wrong.
safe time is used for one purpose - serializable reading from replicas
raft has the same concept to ensure replicas is up to date - commit index.
but we can't use commit index directly, because it is not associated with timestamp.
so we just extend the protocol to assign the entry with motononic growing timestamp in addition to commit index, which is done in the safe way when an entry enters raft pipeline.
other protocols (like Hermes for example) don't require safe timestamp, so it is marked as Nullable in command interface.

Reply:
There are neither mentions of ST in raft paper nor in JRAFT implementation. Are you aware about any other raft implementation besides our "jraft-fork" that has ST concept inside?

Reply:
This is not related to a "specific" raft implementation.
This is about the concept.
applied index and safe timestamp serve the same purpose.
However raft itselft does not require timestamp aware apply index.

Reply:
> Consider this our own mvcc related extention, which is isolated in a replication layer.

I totally agree with this and thus I believe that having raft as safe time unaware module is a better option. It wasn't true even before your PR because of patching. However currently we put the logic much deeper inside raft: all that timeoutNow specific processing, etc.

Unfortunately, it's pretty obvious that time for that discussion has already gone.

Reply:
I believe this is the best solution in the current circumstances. 
PartitionReplicaListener code became much simpler and performs better.
Performance is top priority.
While messing with raft internals generally better be avoided, introduced changes are ultra lightweigh, so I don't see any issues here.